File: apps/qrComm/src/main/java/jp/oist/abcvlib/qr_comm/MainActivity.java
package jp.oist.abcvlib.qr_comm;

import android.graphics.Bitmap;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;

import jp.oist.abcvlib.core.AbcvlibActivity;
import jp.oist.abcvlib.core.AbcvlibLooper;
import jp.oist.abcvlib.core.IOReadyListener;
import jp.oist.abcvlib.core.inputs.PublisherManager;
import jp.oist.abcvlib.core.inputs.phone.ImageData;
import jp.oist.abcvlib.core.inputs.phone.ImageDataSubscriber;
import jp.oist.abcvlib.core.inputs.phone.QRCodeData;
import jp.oist.abcvlib.core.inputs.phone.QRCodeDataSubscriber;
import jp.oist.abcvlib.util.QRCode;

/**
 * Android application showing connection to IOIOBoard, Hubee Wheels, and Android Sensors
 * Initializes socket connection with external python server
 * Runs PID controller locally on Android, but takes PID parameters from python GUI
 * @author Christopher Buckley https://github.com/topherbuckley
 */
public class MainActivity extends AbcvlibActivity implements IOReadyListener, ImageDataSubscriber, QRCodeDataSubscriber {

    private Button showQRCode;
    private boolean isQRDisplayed = false;
    private QRCode qrCode;
    private ImageData imageData;

    public MainActivity() {
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setIoReadyListener(this);

        // Passes Android App information up to parent classes for various usages. Do not modify
        super.onCreate(savedInstanceState);

        // Setup Android GUI. Point this method to your main activity xml file or corresponding int
        // ID within the R class
        setContentView(R.layout.activity_main);

        showQRCode = findViewById(R.id.show_qr_button);
        showQRCode.setOnClickListener(qrCodeButtonClickListener);

        // create a new QRCode object with input args point to the FragmentManager and your layout fragment where you want to generate the qrcode image.
        qrCode = new QRCode(getSupportFragmentManager(), R.id.qrFragmentView);
    }

    @Override
    public void onIOReady(AbcvlibLooper abcvlibLooper) {
        PublisherManager publisherManager = new PublisherManager();
        imageData = new ImageData.Builder(this, publisherManager, this)
                .setPreviewView(findViewById(jp.oist.abcvlib.core.R.id.preview_view))
                .build();
        imageData.addSubscriber(this);

        QRCodeData qrCodeData = new QRCodeData.Builder(this, publisherManager, this).build();
        qrCodeData.addSubscriber(this);

        publisherManager.initializePublishers();
        publisherManager.startPublishers();
    }

    private final View.OnClickListener qrCodeButtonClickListener = v -> {
        if (!isQRDisplayed) {
            // generate new qrcode using the string you want to encode. Use JSONObject.toString for more complex data sets.
            qrCode.generate("Hello World!");
            showQRCode.setText(R.string.back_button_text);
            isQRDisplayed = true;
            imageData.setQrcodescanning(false);
        } else {
            // removes the fragment that holds the last generated qrcode.
            qrCode.close();
            isQRDisplayed = false;
            showQRCode.setText(R.string.qr_button_show);
            imageData.setQrcodescanning(true);
        }
    };

    @Override
    public void onImageDataUpdate(long timestamp, int width, int height, Bitmap bitmap, String qrDecodedData) {
    }

    @Override
    public void onQRCodeDetected(String qrDataDecoded) {
        if (!qrDataDecoded.equals("")){
            Log.i("qrcode", "QR Code Found and decoded: " + qrDataDecoded);
        }
    }
}

File: apps/basicQRTransmitter/src/main/java/jp/oist/abcvlib/basicqrtransmitter/MainActivity.kt
package jp.oist.abcvlib.basicqrtransmitter

import android.os.Bundle
import android.widget.TextView
import jp.oist.abcvlib.core.AbcvlibActivity
import jp.oist.abcvlib.core.inputs.PublisherManager
import jp.oist.abcvlib.core.inputs.phone.QRCodeData
import jp.oist.abcvlib.core.inputs.phone.QRCodeDataSubscriber
import jp.oist.abcvlib.util.Logger
import jp.oist.abcvlib.util.ProcessPriorityThreadFactory
import jp.oist.abcvlib.util.QRCode
import jp.oist.abcvlib.util.ScheduledExecutorServiceWithException
import jp.oist.abcvlib.util.SerialCommManager
import jp.oist.abcvlib.util.SerialReadyListener
import jp.oist.abcvlib.util.UsbSerial
import java.util.concurrent.TimeUnit

/**
 * Android application showing connection to the robot hardware, wheels, and Android sensors
 * Initializes socket connection with external python server
 * Runs PID controller locally on Android, but takes PID parameters from python GUI
 * @author Christopher Buckley https://github.com/topherbuckley
 */
class MainActivity : AbcvlibActivity(), SerialReadyListener, QRCodeDataSubscriber {
    private lateinit var qrCode: QRCode
    private lateinit var publisherManager: PublisherManager
    private lateinit var emojiTextView: TextView

    private var speedL = 0f
    private var speedR = 0f
    private val speed = 0.4f

    private enum class ACTIONS { FORWARD_1, BACKWARD_1, TURN_RIGHT, FORWARD_2, BACKWARD_2, TURN_LEFT }
    private var action = ACTIONS.FORWARD_1

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        qrCode = QRCode(supportFragmentManager, R.id.qrFragmentView)

        emojiTextView = findViewById(R.id.emojiTextView)
        emojiTextView.text = "😊"
    }

    private val swapAction: Runnable = Runnable {
        when (action) {
            ACTIONS.FORWARD_1 -> {
                action = ACTIONS.BACKWARD_1
                goForward()
                updateEmoji("😁")
            }
            ACTIONS.BACKWARD_1 -> {
                action = ACTIONS.TURN_RIGHT
                goBackward()
                updateEmoji("😅")
            }
            ACTIONS.TURN_RIGHT -> {
                action = ACTIONS.FORWARD_2
                turnRight()
                updateEmoji("😎")
            }
            ACTIONS.FORWARD_2 -> {
                action = ACTIONS.BACKWARD_2
                goForward()
                updateEmoji("😁")
            }
            ACTIONS.BACKWARD_2 -> {
                action = ACTIONS.TURN_LEFT
                goBackward()
                updateEmoji("😅")
            }
            ACTIONS.TURN_LEFT -> {
                action = ACTIONS.FORWARD_1
                turnLeft()
                updateEmoji("🤪")
            }
        }
    }

    override fun onSerialReady(usbSerial: UsbSerial) {
        publisherManager = PublisherManager()

        val qrCodeData = QRCodeData.Builder(this, publisherManager, this).build()
        qrCodeData.addSubscriber(this)

        publisherManager.initializePublishers()
        publisherManager.startPublishers()

        setSerialCommManager(SerialCommManager(usbSerial))
        super.onSerialReady(usbSerial)
    }

    public override fun onOutputsReady() {
        publisherManager.initializePublishers()
        publisherManager.startPublishers()
        val executor = ScheduledExecutorServiceWithException(1, ProcessPriorityThreadFactory(Thread.MIN_PRIORITY, "ActionSelector"))
        executor.scheduleAtFixedRate(swapAction, 0, 5, TimeUnit.SECONDS)
    }

    override fun onQRCodeDetected(qrDataDecoded: String) {
        if (qrDataDecoded.isNotEmpty()) {
            Logger.i("qrcode", "QR Code Found and decoded: $qrDataDecoded")
        }
    }

    override fun abcvlibMainLoop() {
        outputs.setWheelOutput(speedL, speedR, false, false)
    }

    private fun goForward() {
        speedL = speed
        speedR = speed
    }

    private fun goBackward() {
        speedL = -speed
        speedR = -speed
    }

    private fun turnRight() {
        speedL = -speed
        speedR = speed
    }

    private fun turnLeft() {
        speedL = speed
        speedR = -speed
    }

    private fun updateEmoji(emoji: String) {
        qrCode.close()
        qrCode.generate(emoji)
        runOnUiThread {
            emojiTextView.text = emoji
        }
    }
}